### DISCUSSION TITLE
Unable to create custom texture reference using an image

### DISCUSSION THREAD

**Discussion by itsron143** [ci-group/revolve2#556] [Ask for help]
Hi,

I'm trying to create a sand-like texture using a `sand.png` as the reference image. 

(These additions can be found [here](https://github.com/itsron143/revolve2/commit/caacf31b6f5fc3789c13822e0a5d8740c86544c5))

Following the implementations of `textures/_checker.py` and `textures/_.gradient.py`, I created `textures/._sand.py`:
```python
from dataclasses import dataclass, field

from revolve2.simulation.scene.geometry.textures import Texture, TextureReference


@dataclass(kw_only=True, frozen=True)
class Sand(Texture):
    """A sand-like texture using a image as reference"""

    reference: TextureReference = field(
        default_factory=lambda: TextureReference(
            content_type="image/png",
            file="./sand.png"
        )
    )

    repeat: tuple[int, int] = field(default=(100, 100))
```

I'm using this with a `sand.png` file to create a custom terrain with a modular robot similar to `examples/1b_custom_terrain`:

```python
def make_custom_sand_terrain() -> Terrain:
    """
    Create a custom terrain with a sand.png as the texture reference.

    :returns: The created terrain.
    """
    return Terrain(
        static_geometry=[
            GeometryPlane(
                pose=Pose(position=Vector3(), orientation=Quaternion()),
                mass=0.0,
                size=Vector3([20.0, 20.0, 0.0]),
                texture=Sand(map_type=MapType.MAP2D,
                             reference=TextureReference(
                                 content_type="image/png",
                                 file="./sand.png")),
            ),
        ]
    )
```

However, when I try to run this custom example I get the following error:
> File "/Users/rounakvyas/.pyenv/versions/revolve/lib/python3.10/site-packages/dm_control/mjcf/element.py", line 534, in _check_valid_attribute
    raise AttributeError( AttributeError: 'content_type' is not a valid attribute for texture )

The complete error stack trace can be found [here](https://pastebin.com/AiUpXr9G). 

From what I understand, `content_type` is an attribute of `TextureReference` and not `Texture`, so I'm unable to understand how to resolve this. 

All my additions to the code can be found [here](https://github.com/itsron143/revolve2/commit/caacf31b6f5fc3789c13822e0a5d8740c86544c5).

Any help will be really appreciated! 

Thanks!

**Comment by oliverweissl** [comment]
Please look at the mujoco xml documentation,
there you see what attributes you need.

https://mujoco.readthedocs.io/en/2.3.6/XMLreference.html#asset-texture

**Reply by itsron143** [reply]
@oliverweissl Removing the `content_type` helped with this issue, thanks! 

However, it still does not render the texture. When I pass the absolute file path for the reference image, the generated xml for Mujoco updates the file name and states that it does not identify the resource.

I pass the full path of the image file:
```python
texture=Sand(map_type=MapType.MAP2D,
        reference=TextureReference(
        file='/Users/rounakvyas/Desktop/Work/VU/msc_thesis/playground/revolve2/_playground/terrains/sand.png')),
```
and I get the following error:
```
  File "/Users/rounakvyas/Desktop/Work/VU/msc_thesis/playground/revolve2/simulators/mujoco_simulator/revolve2/simulators/mujoco_simulator/_simulat
e_scene.py", line 53, in simulate_scene
    model, mapping = scene_to_model(
  File "/Users/rounakvyas/Desktop/Work/VU/msc_thesis/playground/revolve2/simulators/mujoco_simulator/revolve2/simulators/mujoco_simulator/_scene_t
o_model.py", line 138, in scene_to_model
    model = mujoco.MjModel.from_xml_string(xml)
ValueError: Error: resource not found via provider or OS filesystem: 'sand-2bc2fcc6eadee78ac261056a9fa5524812025ec3.png'
Object name = heightmap_0_texture, id = 0, line = 13, column = -1
```

When I check the xml file generated, I see that the `sand.png` file was updated to `sand-2bc2fcc6eadee78ac261056a9fa5524812025ec3.png`

```xml
<mujoco model="scene">
  <compiler autolimits="true" angle="radian"/>
  <option timestep="0.001" gravity="0 0 -9.8100000000000005" integrator="RK4"/>
  <visual>
    <headlight active="0"/>
  </visual>
  <default>
    <default class="/"/>
    <default class="mbs0/"/>
    <default class="mbs1/"/>
  </default>
  <asset>
    <texture name="heightmap_0_texture" type="2d" file="sand-2bc2fcc6eadee78ac261056a9fa5524812025ec3.png" rgb1="0 0 0" rgb2="0 0 0" width="100" h
eight="100"/>
```

It looks like revolve finds the filepath but then the file name gets updated and then mujoco throws a resource not found value error.

**Reply by itsron143** [reply]
Update: This turned out to be a bug in Mujoco itself :) 
https://github.com/google-deepmind/mujoco/issues/1685
